### PR TITLE
Build for Linux ARM32 (armhf) and ARM64 (aarch64)

### PR DESCRIPTION
**Pull request by mattleibow**
**Description of Change**

Add a new Linux architecture: armhf

This is build on Ubuntu 16.04 using the ARM prebuilt toolchain: https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-a/downloads

Using GCC because of: https://github.com/mono/SkiaSharp/pull/1382#discussion_r449709079

**Test Binaries**

Built using Clang: 
 - ARM 32 v7a: [native_linux_arm_linux.zip](https://github.com/mono/SkiaSharp/files/4874470/native_linux_arm_linux.zip)
 - ARM 64 v8: [native_linux_arm64_linux.zip](https://github.com/mono/SkiaSharp/files/4874471/native_linux_arm64_linux.zip)

Built using GCC:
 - ARM 32 v7a: [native_linux_arm_gcc_linux.zip](https://github.com/mono/SkiaSharp/files/4874474/native_linux_arm_gcc_linux.zip)
 - ARM 64 v8: [native_linux_arm64_gcc_linux.zip](https://github.com/mono/SkiaSharp/files/4874476/native_linux_arm64_gcc_linux.zip)

**Test NuGet**

All the packages are on the preview feed as `v2.80.0-pr.1382.31`: `https://aka.ms/skiasharp-eap/index.json`

**Bugs Fixed**

- Related to issue #453

### CODE REVIEW

File: native/linux-arm-clang/build.cake [hunk 1]
```diff
@@ -0,0 +1,58 @@
+DirectoryPath ROOT_PATH = MakeAbsolute(Directory("../.."));
```
Comment (mattleibow) [inline native/linux-arm-clang/build.cake:1]: This script is not complete nor does it work.

Seems libjpeg-turbo is having some issues:
 - https://github.com/libjpeg-turbo/libjpeg-turbo/issues/440
 - https://groups.google.com/g/skia-discuss/c/ZG784t39JyU
 - https://groups.google.com/g/skia-discuss/c/SGA9UVxjabc

File: scripts/azure-pipelines.yml [hunk 2]
```diff
@@ -249,6 +249,14 @@ stages:
...
           target: externals-linux
           additionalArgs: --buildarch=x64
+      - template: azure-templates-bootstrapper.yml # Build Native Linux|armhf (Linux)
+        parameters:
+          name: native_linux_armhf_linux
+          displayName: Build Native Linux|armhf (Linux)
+          vmImage: $(VM_IMAGE_LINUX)
+          docker: scripts/Docker/ubuntu16/armhf
+          target: externals-linux
+          additionalArgs: --buildarch=arm --gnArgs="extra_cflags+=[ '-Wno-psabi' ]"
```
Comment (mattleibow) [inline scripts/azure-pipelines.yml:259]: The extra args are for this issue:
 - https://groups.google.com/g/skia-discuss/c/b29ZStY3Kak
 - https://gcc.gnu.org/gcc-7/changes.html

File: scripts/azure-pipelines.yml [hunk 3]
```diff
@@ -249,6 +249,23 @@ stages:
...
+          docker: scripts/Docker/ubuntu16/gcc-cross
+          target: externals-linux
+          additionalArgs: --buildarch=arm --gnArgs="\"extra_cflags+=[ \'-Wno-psabi\' ]\""
+      - template: azure-templates-bootstrapper.yml # Build Native Linux|arm64 (Linux)
+        parameters:
+          name: native_linux_arm64_linux
+          displayName: Build Native Linux|arm64 (Linux)
+          vmImage: $(VM_IMAGE_LINUX)
+          docker: scripts/Docker/ubuntu16/gcc-cross
+          dockerArgs: --build-arg TOOLCHAIN_ARCH=aarch64-none-linux-gnu --build-arg FONTCONFIG_ARCH=arm64
```
Comment (mattleibow) [inline scripts/azure-pipelines.yml:266]: arm64 or aarch64 images are built using --build-arg

File: scripts/azure-templates-bootstrapper.yml [hunk 4]
```diff
@@ -159,7 +160,7 @@ jobs:
               JavaSdkDirectory: $(JAVA_HOME)
             displayName: Run the bootstrapper for ${{ parameters.target }}
       - ${{ if ne(parameters.docker, '') }}:
-        - bash: docker build --tag skiasharp .
+        - bash: docker build --tag skiasharp ${{ parmeters.dockerArgs }} .
```
Comment (jp2masa) [inline scripts/azure-templates-bootstrapper.yml:163]:
```suggestion
        - bash: docker build --tag skiasharp ${{ parameters.dockerArgs }} .
```

Comment (mattleibow) [reply scripts/azure-templates-bootstrapper.yml]: Thanks!

**Comment by jmacato** [comment]
@mattleibow the Linux ARM64 bins are almost working, though my app had this error:
```
Unhandled exception. System.EntryPointNotFoundException: Unable to find an entry point named 'sk_paint_set_text_encoding' in shared library 'libSkiaSharp'.
   at SkiaSharp.SkiaApi.sk_paint_set_text_encoding(IntPtr param0, SKTextEncoding param1)
   at Avalonia.Skia.FormattedTextImpl..ctor(String text, Typeface typeface, Double fontSize, TextAlignment textAlignment, TextWrapping wrapping, Size constraint, IReadOnlyList`1 spans)
   at Avalonia.Skia.PlatformRenderInterface.CreateFormattedText(String text, Typeface typeface, Double fontSize, TextAlignment textAlignment, TextWrapping wrapping, Size constraint, IReadOnlyList`1 spans)
   at Avalonia.Media.FormattedText.get_PlatformImpl()
   at Avalonia.Media.FormattedText.get_Bounds()
```

@Gillibald should we perhaps update SkiaSharp in Avalonia?

**Comment by Gillibald** [comment]
@jmacato This is targeting the next release. We should be able to update.

**Comment by jmacato** [comment]
@Gillibald awesome :)) well... guess i cant test stuff out without us updating

**Comment by Gillibald** [comment]
The new version is a 1:1 replacement just bump the version number. We can just create a PR that does that and you could use the CI build.

**Comment by jmacato** [comment]
@Gillibald yeah please do, so we can check other issues here just in case

**Comment by mattleibow** [comment]
@jmacato what happens if you bump the version of SkiaSharp in your project to the latest preview on NuGet. Unless Avalonia is p/invoking directly, it should work. The versions should be API compatible as I made no breaks in the .NET API.

**Comment by jmacato** [comment]
@mattleibow i'll try the preview releases first

**Comment by Gillibald** [comment]
I am in the process of creating a PR that changes Avalonia to use the preview. Then you just have to copy the ARM bits manually.

**Comment by jmacato** [comment]
@Gillibald alright i'll wait for that instead :D

**Comment by jmacato** [comment]
Well well well :D 
Screenshot_2020-07-04_19-52-56

Thank you so much @mattleibow & @Gillibald

**Comment by mattleibow** [comment]
CI just failed on the emscripten lols. Why now? Who knows. Maybe a rebuild will fix it?

**Comment by mattleibow** [comment]
Hopefully #1387 will fix it...

**Comment by dinonovak** [comment]
any change to build this for v2.80.2 ?

**Comment by mattleibow** [comment]
This is already out on the SkiaSharp.NativeAssets.Linux packages

**Comment by dinonovak** [comment]
Foun it thax, sorry for confusion

**Comment by mattleibow** [comment]
No prob. Let me know if there are any problems by opening a new issue. Have fun drawing!

**Pull request merged**